Fix enforcing that only one recovery gets retried and succeeds

arteam opened the pull request:
~~#90994 fixed the order or arguments in TransientReceiveRejected and it highlighted the fact that `SingleStartEnforcer` had never actually worked. We never recieve `PeerRecoveryTargetService.Actions.FINALIZE`, so `finalizeReceived` was never true. It was set to true only due to mixed arguments order in `TransientReceiveRejected` which yielded the check useless.~~

We have to register an explicit handler for an `PeerRecoveryTargetService.Actions.FINALIZE` event in order to `finalizeReceived` to be set.

Fixes #91087 
Fixes #91041

arteam committed:
Remove broken SingleStartEnforcer for recovery tests

#90994 fixed the order or arguments in TransientReceiveRejected and it highlighted
the fact that `SingleStartEnforcer` had never actually worked. We never recieve
`PeerRecoveryTargetService.Actions.FINALIZE`, so `finalizeReceived` was never true.
It was set to true only due to mixed arguments order in `TransientReceiveRejected`
which yielded the check useless.

Fixes #91087, #91041

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

arteam committed:
Run spotless

elasticmachine committed:
Merge branch 'main' into fix-index-recovery-tests

arteam commented:
@elasticmachine update elasticsearch-ci/part-1

arteam commented:
@elasticmachine run elasticsearch-ci/part-1

fcofdez commented on test/framework/src/main/java/org/elasticsearch/indices/recovery/AbstractIndexRecoveryIntegTestCase.java:
```diff
@@ -160,14 +158,10 @@ protected void checkTransientErrorsDuringRecoveryAreRetried(String recoveryActio
-            validator.accept(action, request);
-            connection.sendRequest(requestId, action, request, options);
-        });
+        redTransportService.addSendBehavior(
```
I don't think this is the right fix, what's happening here is the following:
- We throw/drop the connection during one of the recovery steps 
    - In this case we schedule a retry after the configured retry delay
- Additionally, we drop the connection from the target -> source
    - When this is detected we try to reestablish the recovery, this takes a while
- Concurrently, while we try to reestablish the recovery connection, the original request that failed makes progress
- The recovery ends, removing it from the on going recoveries collection in the source node
- When the target node reaches the source node trying to reestablish the recovery, this is gone
- The behaviour here is to re-start the recovery, hence the second START_RECOVERY message that we see producing the failure
- The issue is that we set the finalizeReceived value in `TransientReceiveRejected` and that's incorrect, since that's configured to be called only when `recoveryActionToBlock` is received. 

I think that the right fix is something like:
```diff
diff --git a/test/framework/src/main/java/org/elasticsearch/indices/recovery/AbstractIndexRecoveryIntegTestCase.java b/test/framework/src/main/java/org/elasticsearch/indices/recovery/AbstractIndexRecoveryIntegTestCase.java
index 02d46839b29..96f1cc70ac5 100644
--- a/test/framework/src/main/java/org/elasticsearch/indices/recovery/AbstractIndexRecoveryIntegTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/indices/recovery/AbstractIndexRecoveryIntegTestCase.java
@@ -181,9 +181,12 @@ public abstract class AbstractIndexRecoveryIntegTestCase extends ESIntegTestCase
         TransientReceiveRejected handlingBehavior = new TransientReceiveRejected(
             recoveryActionToBlock,
             recoveryStarted,
-            finalizeReceived,
             connectionBreaker
         );
+        redTransportService.addRequestHandlingBehavior(PeerRecoveryTargetService.Actions.FINALIZE, (handler, request, channel, task) -> {
+            finalizeReceived.set(true);
+            handler.messageReceived(request, channel, task);
+        });
         redTransportService.addRequestHandlingBehavior(recoveryActionToBlock, handlingBehavior);

         try {
@@ -580,19 +583,16 @@ public abstract class AbstractIndexRecoveryIntegTestCase extends ESIntegTestCase

         private final String actionName;
         private final AtomicBoolean recoveryStarted;
-        private final AtomicBoolean finalizeReceived;
         private final Runnable connectionBreaker;
         private final AtomicInteger blocksRemaining;

         private TransientReceiveRejected(
             String actionName,
             AtomicBoolean recoveryStarted,
-            AtomicBoolean finalizeReceived,
             Runnable connectionBreaker
         ) {
             this.actionName = actionName;
             this.recoveryStarted = recoveryStarted;
-            this.finalizeReceived = finalizeReceived;
             this.connectionBreaker = connectionBreaker;
             this.blocksRemaining = new AtomicInteger(randomIntBetween(1, 3));
         }
@@ -605,9 +605,6 @@ public abstract class AbstractIndexRecoveryIntegTestCase extends ESIntegTestCase
             Task task
         ) throws Exception {
             recoveryStarted.set(true);
-            if (actionName.equals(PeerRecoveryTargetService.Actions.FINALIZE)) {
-                finalizeReceived.set(true);
-            }
             if (blocksRemaining.getAndUpdate(i -> i == 0 ? 0 : i - 1) != 0) {
                 String rejected = "rejected";
                 String circuit = "circuit";
```
Maybe we can refine it further and assert that we only see the `START_RECOVERY` message once per recovery id.

arteam replied:
Thank you @fcofdez! I noticed that the `PeerRecoveryTargetService.Actions.FINALIZE` event never happened in the handler  `finalizeReceived` wasn't being set when I was debugging the issue, but I decided to restore the state of things before #90994. Fixing the handler is indeed a much better solution!

arteam committed:
Merge remote-tracking branch 'origin/main' into fix-index-recovery-tests

arteam committed:
Revert "Run spotless"

This reverts commit 9fa91214bdee1f659ae9a5fe3d947ee6d14c7f21.

arteam committed:
Revert "Remove broken SingleStartEnforcer for recovery tests"

This reverts commit 69e4ae29336c612fd5164a2ce15ab06c132461de.

arteam committed:
Fix setting finalizeReceived variable

We should register an explicit handler for PeerRecoveryTargetService.Actions.FINALIZE

arteam committed:
Run spotless

fcofdez reviewed (approved):
LGTM 👍 thanks for fixing this!
